File: run-tests.py
# Uploading reports to GitHub

credentials = { }

def read_credentials():
    import os, yaml
    global credentials
    try:
        path = os.path.expanduser("~/credentials.yaml")
        with open(path) as credsfile:
            creds = yaml.safe_load(credsfile.read())
            if creds:
                credentials = creds
    except:
        pass

def ssh_info(job):
    info = credentials[job]
    if info is None:
        return None
    repo = info["repo"]
    key = info["key"]
    url = "ssh://git@github.com/%s" % repo
    return (url, key)

def mkpath(*args):
    import os
    return os.path.join(*args)

def git(*subcmd):
    import subprocess
    cmd = [ "git" ]
    subcmd = list(subcmd)
    if subcmd[0].startswith("!"):
        subcmd[0] = subcmd[0][1:]
    elif subcmd[0] != "-C":
        cmd.extend(["-C", "report"])
    cmd.extend(subcmd)
    result = subprocess.run(cmd,
        stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    return result

def clone(url):
    import os, shutil
    if os.path.exists("report"):
        git("remote", "set-url", "origin", url)
        git("pull", "--ff-only", "origin", "master")
    else:
        shutil.rmtree("report.tmp", ignore_errors=True)
        git("!clone", url, "report.tmp")
        git("-C", "report.tmp", "config",
            "--local", "user.email", "oscar@computeralgebra.de")
        git("-C", "report.tmp", "config",
            "--local", "user.name", "OSCAR Automation")
        git("-C", "report.tmp", "reset", "--hard", "master")
        shutil.move("report.tmp", "report")

def push(url):
    git("remote", "set-url", "origin", url)
    git("push", "-f", "origin", "master")

def add_file(path, contents):
    import os
    dir = os.path.dirname(path)
    if dir != "":
        os.makedirs(mkpath("report", dir), exist_ok=True)
    with open(mkpath("report", path), "w") as outfile:
        outfile.write(contents)
    git("add", path)

def commit(msg):
    git("commit", "-m", msg)

def upload(job, build, files):
    import os
    read_credentials()
    if job not in credentials:
        return
    url, key = ssh_info(job)
    keyfile = os.path.expanduser("~/.ssh_git_key." + job)
    with open(keyfile, "w") as keyfp:
      keyfp.write(key)
    os.chmod(keyfile, 0o600)
    os.environ["GIT_SSH_COMMAND"] = "ssh -o 'StrictHostKeyChecking no' -i " + keyfile
    clone(url)
    for path, contents in files.items():
        add_file(path, contents)
    commit("Build " + str(build))
    push(url)
    try: os.remove(keyfile)
    except: pass

# Test runner

default_timeout = 1800

class YamlDB(dict):
    def __init__(self, path):
        import yaml, fcntl
        super(YamlDB, self).__init__()
        self.path = path
        self.lock = open("%s.lock" % path, "w+")
        fcntl.lockf(self.lock, fcntl.LOCK_EX)
        try:
            with open(path, "r") as fp:
                self.update(yaml.safe_load(fp))
        except FileNotFoundError:
            pass
    def sync(self):
        import os, yaml
        tmppath = "%s.tmp" % self.path
        data = {}
        data.update(self)
        with open(tmppath, "w") as fp:
            fp.write(yaml.safe_dump(data))
        os.rename(tmppath, self.path)
    def _unlock(self):
        import fcntl
        fcntl.lockf(self.lock, fcntl.LOCK_UN)
        self.lock = None
    def close(self):
        self.sync()
        self._unlock()
    def __del__(self):
        if self.lock is not None:
            self._unlock()

def load_config():
    import yaml
    with open("meta/tests/config.yaml") as config:
        return yaml.safe_load(config.read())

def make_job_url(buildnum):
    import os
    url = os.environ["JENKINS_URL"]
    job = os.environ["JOB_NAME"]
    if not url.endswith("/"): url += "/"
    url += "job/%s/%s/" % (job, buildnum)
    return url

class TestRunner:
    def __init__(self):
        # Init instance variables
        import os, threading, concurrent.futures as futures
        self.failed_tests = False
        self.buildnum = os.environ.get("BUILD_NUMBER", "0")
        self.jenkins_home = os.environ["JENKINS_HOME"]
        self.job = os.environ["JOB_NAME"]
        self.build_url = os.environ["BUILD_URL"]
        self.maxjobs = int(os.environ.get("BUILDJOBS", 4))
        if not self.build_url.endswith("/"): self.build_url += "/"
        self.log_url = self.build_url + "artifact/logs/build-" + self.buildnum
        self.logdir = "logs/build-%s" % self.buildnum
        self.start_date = None
        self.end_date = None
        self.successes = []
        self.failures = []
        self.testresults = []
        self.lock = threading.Lock()
        self.threadpool = futures.ThreadPoolExecutor(max_workers=self.maxjobs)
        self.futures = {}
        # Create log dir for this build
        try: os.makedirs(self.logdir, exist_ok=True)
        except: pass
    def report(self, msg, index, exitcode):
        if exitcode == 0:
            self.successes.append((index, msg))
        else:
            self.failures.append((index, msg))
    def _run(self, test, index = 0):
        import os, subprocess, datetime, yaml, re
        # Retrieve information about this test.
        testscript = "meta/tests/" + test["script"]
        testname = test["name"]
        testfilename = re.sub("[^-._a-zA-Z0-9]+", "-", testname)
        info = {}
        info["name"] = testname
        if "timeout" in test:
            timeout = test["timeout"]
        else:
            timeout = default_timeout
        logfile = os.path.join(self.logdir, testfilename + ".log")
        info["log"] = "%s/%s.log" % (self.log_url, testfilename)
        def log(s):
            with open(logfile, "ab") as logfp:
                if type(s) == type(u""):
                    s = bytearray(s, "utf-8")
                logfp.write(s)
        # Run the test and record its status
        try:
            start_time = datetime.datetime.now()
            with self.lock:
                if self.start_date is None:
                    self.start_date = start_time.strftime("%Y-%m-%d")
            start = start_time.strftime("%Y-%m-%d %H:%M")
            info["start"] = start
            start_short = start_time.strftime("%H:%M")
            log("=== %s (%s) at %s\n" % (testname, testscript, start))
            cmd = testscript
            cmd += " >>" + logfile + " 2>&1"
            result = subprocess.run(cmd, shell=True, timeout = timeout)
            exitcode = result.returncode
            if exitcode == 0:
                verbose_status = "SUCCESS"
                statuscode = "\u2705"
            else:
                verbose_status = "FAILURE"
                statuscode = "\u274C"
        except subprocess.TimeoutExpired as result:
            output = result.stdout
            exitcode = -1
            verbose_status = "TIMEOUT"
            statuscode = "\u26A0"
        except:
            output = "INTERNAL ERROR\n"
            verbose_status = "INTERNAL ERROR"
            statuscode = "\u2049"
            exitcode = -1
        with self.lock:
            self.failed_tests = self.failed_tests or exitcode != 0
        info["success"] = (exitcode == 0)
        info["exitcode"] = exitcode
        info["status"] = verbose_status.lower()
        if verbose_status == "FAILURE":
            verbose_status += " (status = %d)" % exitcode
        # Timing information
        stop_time = datetime.datetime.now()
        stop = stop_time.strftime("%Y-%m-%d %H:%M")
        with self.lock:
            new_end_date = stop_time.strftime("%Y-%m-%d")
            if self.end_date is None or new_end_date > self.end_date:
                self.end_date = new_end_date
        duration = (stop_time - start_time).seconds
        info["duration"] = duration
        # Make a persistent record of the last success/first failure
        jobstate_dir = self.jenkins_home + "/jobstate"
        jobstate = mkpath(jobstate_dir, self.job + ".yaml")
        os.makedirs(jobstate_dir, exist_ok=True)
        with self.lock:
            db = YamlDB(jobstate)
            info["last_success"] = False
            info["last_success_url"] = None
            info["first_failure"] = False
            info["first_failure_url"] = None
            if exitcode == 0:
                last_success = ""
                first_failure = ""
                db[testname] = [ self.buildnum, "" ]
            else:
                if testname in db:
                    if db[testname][1] == "":
                        db[testname][1] = self.buildnum
                else:
                    db[testname] = [ "", self.buildnum ]
                if db[testname][0] == "":
                    last_success = "unknown"
                    first_failure = "unknown"
                else:
                    err = db[testname]
                    url = [ make_job_url(str(n)) for n in err ]
                    last_success = "[%s](%s)" % (err[0], url[0])
                    first_failure = "[%s](%s)" % (err[1], url[1])
                    info["last_success"] = int(err[0])
                    info["last_success_url"] = url[0]
                    info["first_failure"] = int(err[1])
                    info["first_failure_url"] = url[1]
            db.close()
        # Log test results
        testsummary = "| %s | %s [%s](%s) | %s | %s | %s | %s |" % \
            (testname, statuscode, verbose_status.lower().capitalize(),
             self.log_url + "/" + testfilename + ".log", start_short,
             "%d seconds" % duration, last_success, first_failure)
        with self.lock:
            self.report(testsummary, index, exitcode)
            self.testresults.append((index, info))
        log("=== %s at %s\n" % (verbose_status, stop))
        return "Testing: %-19s at %s => %s" % (testname, start, verbose_status)
    def run(self, test):
        import sys
        msg = self._run(test)
        print(msg)
        sys.stdout.flush()
    def _start(self, test, index):
        after = test.get("after", None)
        dep = self.futures.get(after, None)
        if after is not None and dep is None:
            print("Warning: Invalid dependency %s => %s" % (test["name"],
                    after))
        else:
            if dep is not None:
                self.futures[after].result()
        return self._run(test, index)
    def start(self, test, index=0):
        with self.lock:
            future = self.threadpool.submit(self._start, test, index)
            self.futures[test["name"]] = future
            return future
    def finish(self):
        import yaml
        def reindex(l):
            l.sort()
            l[:] = [ item for index, item in l ]
        print("Logs: " + self.log_url + "/")
        # Print a human readable report
        report_form = ""
        report_form += "## [Build %s](%s)\n\n" % (self.buildnum, self.build_url)
        report_form += "* Started on: %s\n" % self.start_date
        report_form += "* Ended on: %s\n\n" % self.end_date
        report_form += \
            "| Test Name    | Result | Start | Duration | Last Success | First Failure |\n" + \
            "|:-------------|:-------|:------|:---------|:-------------|:--------------|\n"
        reindex(self.failures)
        reindex(self.successes)
        reindex(self.testresults)
        if len(self.failures) > 0:
            report_form += "\n".join(self.failures)
            report_form += "\n"
        if len(self.successes) > 0:
            report_form += "\n".join(self.successes)
            report_form += "\n"
        # Dump report into a yaml file
        upload(self.job, self.buildnum, {
            "README.md" : report_form,
            "_data/ci.yml" : yaml.dump({
                "build": int(self.buildnum),
                "build_url": self.build_url,
                "job": self.job,
                "tests": self.testresults
            }, default_flow_style=False)
        })

def run_tests(tests):
    import sys
    testrunner = TestRunner()
    tasks = [ testrunner.start(test, index)
        for index, test in zip(range(len(tests)), tests) ]
    for task in tasks:
        print(task.result())
        sys.stdout.flush()
    testrunner.finish()
    if testrunner.failed_tests:
        exit(1)

if __name__ == "__main__":
    tests = load_config()
    run_tests(tests)

File: run-tests.rb
#!/usr/bin/env ruby

require "timeout"
require "yaml"
require "yaml/store"
require "open3"

require_relative "settings"
require_relative "utils"

def parse_mem(data)
  case data
  when Integer, nil then
    data
  when /^([0-9]+)$/ then
    $1.to_i
  when /^([0-9]+)[kK]$/ then
    $1.to_i * 1024
  when /^([0-9]+)[mM]$/ then
    $1.to_i * 1024 * 1024
  when /^([0-9]+)[gG]$/ then
    $1.to_i * 1024 * 1024 * 1024
  else
    nil
  end
end

class GitServer
  def initialize(sshbase, httpsbase, credfile)
    @sshbase = sshbase
    @httpsbase = httpsbase
    @credentials = {}
    begin
      @credentials = YAML.safe_load(File.read(credfile))
    rescue
      File.open(".gitlog", "a") do | fp |
        fp.puts(YAML.dump({
          "time" => Time.now.to_s,
          "info" => "missing or unreadable credentials file",
          "file" => credfile
        }))
      end
    end
  end

  def [](job)
    @credentials[job]
  end

  def repo_info(job)
    info = @credentials[job]
    if not info then
      return nil
    end
    repo = info["repo"]
    key = info["key"]
    sshurl = "#{@sshbase}/#{repo}"
    httpsurl = "#{@httpsbase}/#{repo}"
    return repo, key, sshurl, httpsurl
  end
end

class GitRepo

  attr_reader :job, :path, :sshurl, :httpsurl, :key

  def initialize(server, job)
    @server = server
    @env = { }
    @job = job
    @path, @key, @sshurl, @httpsurl = server.repo_info(job)
  end

  def git(*subcmd, autocd: true)
    cmd = [ "git" ]
    if autocd then
      cmd += [ "-C", "report" ]
    end
    cmd += subcmd
    stdout, stderr, status = Open3.capture3(@env, *cmd, stdin_data: "")
    File.open(".gitlog", "a") do | fp |
      fp.puts(YAML.dump({
        "time" => Time.now.to_s,
        "cmd" => cmd.join(" "),
        "stdout" => stdout,
        "stderr" => stderr,
      }))
    end
    return stdout, stderr, status
  end

  def fetch
    if Dir.exist?("report") then
      git("remote", "set-url", "origin", "--", @sshurl)
      git("pull", "--ff-only", "origin", "master")
    else
      FileUtils.rm_tree("report.tmp")
      git("clone", @sshurl, "report.tmp", autocd: false)
      git("-C", "report.tmp", "config",
        "--local", "user.email", "oscar@computeralgebra.de", autocd: false)
      git("-C", "report.tmp", "config",
        "--local", "user.name", "OSCAR Automation", autocd: false)
      git("-C", "report.tmp", "reset", "--hard", "master")
      FileUtils.mv("report.tmp", "report")
    end
  end

  def push
    git("remote", "set-url", "origin", "--", @sshurl)
    git("push", "-f", "origin", "master")
  end

  def add_file(path, contents)
    dir = File.dirname(path)
    if dir != "" then
      dir = "report/#{dir}"
      FileUtils.mkdir_p(dir)
      File.open("report/#{path}", "wb") do | fp |
        fp.write(contents)
      end
    end
    git("add", "--", path)
  end

  def add_tree(path)
    Dir.glob("**/{*,.*}", base: path) do | relpath |
      loc = File.join(path, relpath)
      add_file relpath, File.read(loc) if File.file?(loc)
    end
  end

  def commit(msg)
    git("commit", "-m", msg)
  end

  def upload(job, msg, files: {}, dirs: [])
    return if not @server[job]
    keyfile = File.expand_path("~/.ssh_git_key.#{job}")
    File.write(keyfile, @key)
    FileUtils.chmod(0o600, keyfile)
    @env = {
      "GIT_SSH_COMMAND" => "ssh -o 'StrictHostKeyChecking no' -i #{keyfile}"
    }
    fetch
    dirs.each do | path |
      add_tree(path)
    end
    files.each do | path, contents |
      add_file(path, contents)
    end
    commit(msg)
    push
    FileUtils.rm_f(keyfile)
  end
end

class Logger
  def initialize(path)
    @path = path
    FileUtils.mkdir_p(File.dirname(@path))
  end

  def <<(msg)
    File.open(@path, "a") do | fp |
      fp.write msg
    end
  end
end

class TestRunner

  attr_reader :failed_tests

  DEFAULT_TIMEOUT = 1800

  def initialize(repo, tests)
    @repo = repo
    @tests = tests
    @failed_tests = false
    @buildnum = ENV["BUILD_NUMBER"] || "0"
    @workspace = $WORKSPACE
    @job = repo.job
    @buildurl = strip_trailing_slashes(ENV["BUILD_URL"])
    @artifacturl = "#{@buildurl}/artifact"
    @jenkinsurl = strip_trailing_slashes(ENV["JENKINS_URL"])
    @maxjobs = (ENV["BUILDJOBS"] || 4).to_i
    @logdir = "logs/build-#{@buildnum}"
    @logurlbase = "#{@artifacturl}/#{@logdir}"
    @startdate = nil
    @enddate = nil
    @successes = {}
    @failures = {}
    @testresults = {}
    @jenkinshome = File.expand_path("#{$WORKSPACE}/../..")
    @jobstatepath = "#{@jenkinshome}/jobstate/#{@job}.yaml"
    FileUtils.mkdir_p(File.dirname(@jobstatepath))
    @jobstate = YAML::Store.new(@jobstatepath)
    @lock = Thread::Mutex.new
    @builddir = Dir.pwd
  end

  class Constraints
    def initialize(tests)
      @constraints = {}
      @all = []
      @blocked = false
      for test in tests do
        @all |= test["uses"].to_s.split if test["uses"] != "*"
        @all |= [ test["name"] ]
      end
    end

    private def get_uses(test)
      result = test["uses"].to_s.split + [ test["name"] ]
      result = @all if result.include?("*")
      return result
    end

    def add?(test)
      uses = get_uses(test)
      result = uses.all? { | use | not @constraints[use] }
      if result then
        uses.each { | use | @constraints[use] = true }
      end
      return result
    end

    def remove(test)
      uses = get_uses(test)
      uses.each { | use | @constraints[use] = false }
    end
  end

  private def run_tests_in_parallel
    todo = Thread::Queue.new
    done = Thread::Queue.new
    threads = []
    @maxjobs.times do
      threads << Thread.new(todo, done) do | todo, done |
        loop do
          test = todo.pop
          run_test(test)
          done.push test
        end
      end
    end
    constraints = Constraints.new(@tests)
    complete = 0
    remaining = @tests.dup
    while not remaining.empty? do
      scheduled = false
      for test in remaining do
        if constraints.add?(test) then
          todo.push test
          remaining.delete test
          scheduled = true
          break
        end
      end
      next if scheduled
      last_test = done.pop
      complete += 1
      constraints.remove(last_test)
    end
    while complete < @tests.size do
      done.pop
      complete += 1
    end
    for thread in threads do
      thread.kill
      thread.join
    end
  end

  private def strip_trailing_slashes(url)
    url = url.dup
    while url.end_with?("/") do url.chop! end
    url
  end

  private def report(name, msg, success:)
    if success then
      @successes[name] = msg
    else
      @failures[name] = msg
    end
  end

  private def buildurl
    "#{@buildurl}/"
  end

  private def make_build_url(buildnum)
    "#{@jenkinsurl}/job/#{@job}/#{buildnum}/"
  end

  private def update_jobstate(info, testname, exitcode)
    last_success = first_failure = nil
    @lock.synchronize do
      @jobstate.transaction do
        info["last_success"] = false
        info["last_success_url"] = nil
        info["first_failure"] = false
        info["first_failure_url"] = nil
        if exitcode.zero? then
          last_success = ""
          first_failure = ""
          @jobstate[testname] = [ @buildnum, "" ]
        else
          if @jobstate[testname] then
            if @jobstate[testname][1] == ""
              @jobstate[testname][1] = @buildnum
            end
          else
            @jobstate[testname] = [ "", @buildnum ]
          end
          if @jobstate[testname][0] == "" then
            last_success = "unknown"
            first_failure = "unknown"
          else
            err = @jobstate[testname]
            url = err.map { | n | make_build_url(n.to_s) }
            last_success = "[#{err[0]}](#{url[0]})"
            first_failure = "[#{err[1]}](#{url[1]})"
            info["last_success"] = err[0].to_i
            info["last_success_url"] = url[0]
            info["first_failure"] = err[1].to_i
            info["first_failure_url"] = url[1]
          end
        end
      end
    end
    return last_success, first_failure
  end

  def wspath(path)
    File.expand_path(path, $WORKSPACE)
  end

  private def test_command(test)
    if test["script"] then
      wspath(%Q{meta/tests/#{test["script"]}})
    elsif test["sh"] then
      test["sh"]
    elsif test["julia"] then
      %w{julia -e} + [ test["julia"] ]
    elsif test["package"] then
      %w{julia -e} + [ %Q{import Pkg; Pkg.test("#{test["package"]}")} ]
    elsif test["gap"] then
      %w{gap --quitonbreak -c} + [ test["gap"] ]
    elsif test["gappkg"] then
      gapcode = %Q{Read(Filename(DirectoriesPackageLibrary("#{test["gappkg"]}", "tst"), "testall.g"));}
      %w{gap --quitonbreak -c} + [ gapcode ]
    elsif test["notebook"] then
      [ "ruby", wspath("meta/check-julia-notebook.rb"),
        wspath(test["notebook"]) ]
    else
      fail "invalid test type" # TODO
    end
  end

  private def run_test(test)
    testname = test["name"]
    testfilename = testname.gsub(/[^-._a-zA-Z0-9]+/, "-")
    testid = testname.gsub(/[^a-zA-Z0-9]+/, "-").downcase
    info = { "name" => testname, "id" => testid }
    timeout = test["timeout"] || DEFAULT_TIMEOUT
    logfile = "#{@logdir}/#{testfilename}.log"
    logurl = info["log"] = "#{@logurlbase}/#{testfilename}.log"
    log = Logger.new(logfile)
    begin
      start_time = Time.now
      @lock.synchronize do
        @start_date ||= start_time.strftime("%Y-%m-%d")
      end
      start = start_time.strftime("%Y-%m-%d %H:%M")
      info["start"] = start
      start_short = start_time.strftime("%H:%M")
      log << "=== #{testname} at #{start_short}\n"
      testcmd = test_command(test)
      juliaenv = "#{$WORKSPACE}/julia-env"
      standalone = test["standalone"]
      extra_pkgs = []
      if standalone then
        testdir = "#{$WORKSPACE}/test-env/#{testfilename}"
        FileUtils.rm_tree(testdir)
        FileUtils.mkdir_p(testdir)
        case standalone
        when String then
          extra_pkgs = standalone.split
        when Array then
          extra_pkgs = standalone
        end
      else
        testdir = "."
      end
      polymake_user_dir = "#{$WORKSPACE}/.polymake/#{testfilename}"
      FileUtils.rm_tree(polymake_user_dir)
      FileUtils.mkdir_p(polymake_user_dir)
      testenv = {
        "POLYMAKE_USER_DIR" => polymake_user_dir
      }
      pid = -1
      status = nil
      Timeout.timeout(timeout) do
        proceed = true
        for pkg in extra_pkgs do
          pid = spawn(testenv,
            "julia -e 'import Pkg; Pkg.activate(\".\"); Pkg.add(\"#{pkg}\")'",
            err: [ :child, :out], out: [ logfile, "a"], chdir: testdir)
          _, status = Process.waitpid2(pid)
          if not status.success? then
            proceed = false
            break
          end
        end
        if proceed then
          pid =
            if testcmd.is_a?(String) then
              spawn(testenv, testcmd,
                err: [ :child, :out ], out: [ logfile, "a" ], pgroup: true,
                chdir: testdir)
            else
              spawn(testenv, [ testcmd.first, testcmd.first ], *testcmd[1..-1],
                err: [ :child, :out ], out: [ logfile, "a" ], pgroup: true,
                chdir: testdir)
            end
          _, status = Process.waitpid2(pid)
        end
      end
      exitcode = status.exitstatus
      if exitcode.zero? then
        verbose_status = "SUCCESS"
        statuscode = "\u2705"
      else
        verbose_status = "FAILURE"
        statuscode = "\u274C"
      end
    rescue Timeout::Error
      if pid >= 0 then
        pgid = Process.getpgid(pid)
        Process.kill("TERM", pgid)
        sleep 5
        Process.kill("KILL", pgid)
      end
      exitcode = -1
      verbose_status = "TIMEOUT"
      statuscode = "\u26A0"
    rescue StandardError => ex
      log << "#{ex.backtrace.join("\n")}\n"
      log << "#{ex.inspect}\n"
      exitcode = -1
      verbose_status = "INTERNAL ERROR"
      statuscode = "\u2049"
    end
    @lock.synchronize do
      @failed_tests ||= exitcode != 0
    end
    info["success"] = exitcode.zero?
    info["exitcode"] = exitcode
    info["status"] = verbose_status.downcase
    if verbose_status == "FAILURE" then
      verbose_status += " (status = #{exitcode})"
    end
    stop_time = Time.now
    stop = stop_time.strftime("%Y-%m-%d %H:%M")
    @lock.synchronize do
      new_end_date = stop_time.strftime("%Y-%m-%d")
      @end_date ||= new_end_date
      @end_date = new_end_date if new_end_date > @end_date
    end
    duration = (stop_time - start_time).round
    info["duration"] = duration
    last_success, first_failure = update_jobstate(info, testname, exitcode)
    @testresults[testname] = info
    testsummary =  "| #{testname} "
    testsummary << "| #{statuscode} [#{verbose_status}](#{logurl}) "
    testsummary << "| #{start_short} | #{duration} seconds "
    testsummary << "| #{last_success} | #{first_failure} "
    testsummary << "|"
    @lock.synchronize do
      report(testname, testsummary, success: exitcode.zero?)
    end
    log << "=== #{verbose_status} at #{stop}\n"
    puts format("Testing: %-19<name>s at %<time>s => %<status>s",
      name: testname, time: start, status: verbose_status)
  end

  def ordered(results)
    result = []
    for test in @tests do
      result << results[test["name"]] if results[test["name"]]
    end
    return result
  end

  def finish_report
    successes = ordered(@successes)
    failures = ordered(@failures)
    report = []
    report << "## [Build #{@buildnum}](#{buildurl})\n\n"
    report << "* Started on: #{@start_date}\n"
    report << "* Ended on: #{@end_date}\n\n"
    report << "| Test Name    | Result | Start | Duration | Last Success | First Failure |\n"
    report << "|:-------------|:-------|:------|:---------|:-------------|:--------------|\n"
    report << "#{failures.join("\n")}\n" unless failures.empty?
    report << "#{successes.join("\n")}\n" unless successes.empty?
    @report = report.join
    puts "Logs: #{@logurlbase}"
  end

  def run_all(parallelize:, memlimit: nil)
    if memlimit then
      Process.setrlimit(:DATA, memlimit, memlimit)
    end
    if parallelize > 1 then
      run_tests_in_parallel
    else
      for test in @tests do
        run_test(test)
      end
    end
    finish_report
    if @repo.path then
      @repo.upload(@job, "Build ##{@buildnum}",
        dirs: [
          "meta/layout",
        ],
        files: {
          "README.md" => @report,
          "_data/ci.yml" => YAML.dump({
            "build" => @buildnum.to_i,
            "build_url" => buildurl,
            "job" => @job,
            "organization" => File.dirname(@repo.path),
            "repo" => File.basename(@repo.path),
            "repourl" => @repo.httpsurl,
            "tests" => ordered(@testresults),
          })
        })
    end
  end

end

def get_par_info(parallelize)
  case parallelize
  when nil, false, true then
    parallelize ? (ENV["BUILDJOBS"] || 4).to_i : 0
  when Integer then
    [ parallelize, 0 ].max
  when Hash then
    value = parallelize
    value ||= (ENV["BUILDJOBS"] || 4).to_i
  end
end

def main
  FileUtils.mkdir_p "logs"
  jobname = ENV["JOB_NAME"]
  config = YAML.safe_load(File.read("meta/tests/config.yaml"))
  if config["jobinfo"] then
    config.update(config["jobinfo"][jobname] || {})
  end
  github = GitServer.new("ssh://git@github.com", "https://github.com",
    ENV["CREDENTIALS"] || "/config/credentials.yaml")
  repo = GitRepo.new(github, jobname)
  tests = config["tests"]
  parallelize = get_par_info(config["parallelize"])
  memlimit = parse_mem(config["memlimit"])
  testrunner = TestRunner.new(repo, tests)
  testrunner.run_all(parallelize: parallelize, memlimit: memlimit)
  exit 1 if testrunner.failed_tests
end

main if caller.empty?
